README.md
<!-- TITLE/ -->

<h1>Feedr</h1>

<!-- /TITLE -->


<!-- BADGES/ -->

<span class="badge-travisci"><a href="http://travis-ci.com/bevry/feedr" title="Check this project's build status on TravisCI"><img src="https://img.shields.io/travis/com/bevry/feedr/master.svg" alt="Travis CI Build Status" /></a></span>
<span class="badge-npmversion"><a href="https://npmjs.org/package/feedr" title="View this project on NPM"><img src="https://img.shields.io/npm/v/feedr.svg" alt="NPM version" /></a></span>
<span class="badge-npmdownloads"><a href="https://npmjs.org/package/feedr" title="View this project on NPM"><img src="https://img.shields.io/npm/dm/feedr.svg" alt="NPM downloads" /></a></span>
<span class="badge-daviddm"><a href="https://david-dm.org/bevry/feedr" title="View the status of this project's dependencies on DavidDM"><img src="https://img.shields.io/david/bevry/feedr.svg" alt="Dependency Status" /></a></span>
<span class="badge-daviddmdev"><a href="https://david-dm.org/bevry/feedr#info=devDependencies" title="View the status of this project's development dependencies on DavidDM"><img src="https://img.shields.io/david/dev/bevry/feedr.svg" alt="Dev Dependency Status" /></a></span>
<br class="badge-separator" />
<span class="badge-githubsponsors"><a href="https://github.com/sponsors/balupton" title="Donate to this project using GitHub Sponsors"><img src="https://img.shields.io/badge/github-donate-yellow.svg" alt="GitHub Sponsors donate button" /></a></span>
<span class="badge-patreon"><a href="https://patreon.com/bevry" title="Donate to this project using Patreon"><img src="https://img.shields.io/badge/patreon-donate-yellow.svg" alt="Patreon donate button" /></a></span>
<span class="badge-flattr"><a href="https://flattr.com/profile/balupton" title="Donate to this project using Flattr"><img src="https://img.shields.io/badge/flattr-donate-yellow.svg" alt="Flattr donate button" /></a></span>
<span class="badge-liberapay"><a href="https://liberapay.com/bevry" title="Donate to this project using Liberapay"><img src="https://img.shields.io/badge/liberapay-donate-yellow.svg" alt="Liberapay donate button" /></a></span>
<span class="badge-buymeacoffee"><a href="https://buymeacoffee.com/balupton" title="Donate to this project using Buy Me A Coffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-yellow.svg" alt="Buy Me A Coffee donate button" /></a></span>
<span class="badge-opencollective"><a href="https://opencollective.com/bevry" title="Donate to this project using Open Collective"><img src="https://img.shields.io/badge/open%20collective-donate-yellow.svg" alt="Open Collective donate button" /></a></span>
<span class="badge-crypto"><a href="https://bevry.me/crypto" title="Donate to this project using Cryptocurrency"><img src="https://img.shields.io/badge/crypto-donate-yellow.svg" alt="crypto donate button" /></a></span>
<span class="badge-paypal"><a href="https://bevry.me/paypal" title="Donate to this project using Paypal"><img src="https://img.shields.io/badge/paypal-donate-yellow.svg" alt="PayPal donate button" /></a></span>
<span class="badge-wishlist"><a href="https://bevry.me/wishlist" title="Buy an item on our wishlist for us"><img src="https://img.shields.io/badge/wishlist-donate-yellow.svg" alt="Wishlist browse button" /></a></span>

<!-- /BADGES -->


<!-- DESCRIPTION/ -->

Use feedr to fetch the data from a remote url, respect its caching, and parse its data. Despite its name, it's not just for feed data but also for all data that you can feed into it (including binary data).

<!-- /DESCRIPTION -->


<!-- INSTALL/ -->

<h2>Install</h2>

<a href="https://npmjs.com" title="npm is a package manager for javascript"><h3>npm</h3></a>
<ul>
<li>Install: <code>npm install --save feedr</code></li>
<li>Require: <code>require('feedr')</code></li>
</ul>

<h3><a href="https://editions.bevry.me" title="Editions are the best way to produce and consume packages you care about.">Editions</a></h3>

<p>This package is published with the following editions:</p>

<ul><li><code>feedr</code> aliases <code>feedr/source/index.js</code></li>
<li><code>feedr/source/index.js</code> is esnext source code with require for modules</li></ul>

<p>Environments older than Node.js v8 may need <a href="https://babeljs.io/docs/usage/polyfill/" title="A polyfill that emulates missing ECMAScript environment features">Babel's Polyfill</a> or something similar.</p>

<!-- /INSTALL -->


## Usage

``` javascript
// Create a new feedr instance
const feedr = require('feedr').create({/* optional configuration */})

// Prepare our feeds that we want read
const feeds = {
	someAtomFeed: 'https://github.com/bevry/feedr/commits/for-testing.atom'
	someJsonFile: 'https://raw.githubusercontent.com/bevry/feedr/for-testing/package.json'
}

// Read a single feed
feedr.readFeed(feeds.someAtomFeed, {/* optional configuration */}, function (err, data, headers) {
	console.log(err, data, headers)
})

// Read all the feeds together
feedr.readFeeds(feeds, {/* optional configuration */}, function (err, result) {
	console.log(err, result.someAtomFeed, result.someJsonFile)
})
```


## Configuration

Feed configuration defaults / global configuration properties are:

- `log` defaults to `null`, log function to use
- `tmpPath` defaults to system tmp path, the tempory path to cache our feedr results to
- `cache` defaults to one day `1000*60*60*24`, available values:
	- `Number` prefers to use the cache when it is within the range of the number in milliseconds
	- `true` prefers to use the cache when the response headers indicate that the cache is still valid
	- `"preferred"` will always use the cache if the cache exists
	- `false` will never use the cache
- `xml2jsOptions` defaults to `null`, the options to send to [xml2js](https://github.com/Leonidas-from-XIV/node-xml2js)
- `requestOptions` defaults to `null`, the options to send to [request](https://github.com/mikeal/request)

Feed configuration properties are:

- `url` required, the url to fetch
- `hash` defaults to hash of the url, the hashed url for caching
- `name` defaults to hash, the name of the feed for use in debugging
- `path` defaults to tmp feed path, the path to save the file to
- `parse` defaults to `true`, whether or not we should attempt to parse the response data, supported values are `xml`, `json`, `cson`, `yaml`, `string`, `raw`/`false`. If `true` will try all the available parsers. Can also be a function with the signature `({response, data, feed, feedr}, next(err, data))`
- `check` defaults to `true`, whether or not we should check the response data for custom error messages. Can also be a function with the signature `({response, data, feed, feedr}, next(err))`
- `xml2jsOptions` defaults to global value, the options to send to [xml2js](https://github.com/Leonidas-from-XIV/node-xml2js)
- `requestOptions` defaults to global value, the options to send to [request](https://github.com/mikeal/request)


<!-- HISTORY/ -->

<h2>History</h2>

<a href="https://github.com/bevry/feedr/blob/master/HISTORY.md#files">Discover the release history by heading on over to the <code>HISTORY.md</code> file.</a>

<!-- /HISTORY -->


<!-- CONTRIBUTE/ -->

<h2>Contribute</h2>

<a href="https://github.com/bevry/feedr/blob/master/CONTRIBUTING.md#files">Discover how you can contribute by heading on over to the <code>CONTRIBUTING.md</code> file.</a>

<!-- /CONTRIBUTE -->


<!-- BACKERS/ -->

<h2>Backers</h2>

<h3>Maintainers</h3>

These amazing people are maintaining this project:

<ul><li><a href="http://balupton.com">Benjamin Lupton</a> — <a href="https://github.com/bevry/feedr/commits?author=balupton" title="View the GitHub contributions of Benjamin Lupton on repository bevry/feedr">view contributions</a></li></ul>

<h3>Sponsors</h3>

No sponsors yet! Will you be the first?

<span class="badge-githubsponsors"><a href="https://github.com/sponsors/balupton" title="Donate to this project using GitHub Sponsors"><img src="https://img.shields.io/badge/github-donate-yellow.svg" alt="GitHub Sponsors donate button" /></a></span>
<span class="badge-patreon"><a href="https://patreon.com/bevry" title="Donate to this project using Patreon"><img src="https://img.shields.io/badge/patreon-donate-yellow.svg" alt="Patreon donate button" /></a></span>
<span class="badge-flattr"><a href="https://flattr.com/profile/balupton" title="Donate to this project using Flattr"><img src="https://img.shields.io/badge/flattr-donate-yellow.svg" alt="Flattr donate button" /></a></span>
<span class="badge-liberapay"><a href="https://liberapay.com/bevry" title="Donate to this project using Liberapay"><img src="https://img.shields.io/badge/liberapay-donate-yellow.svg" alt="Liberapay donate button" /></a></span>
<span class="badge-buymeacoffee"><a href="https://buymeacoffee.com/balupton" title="Donate to this project using Buy Me A Coffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-yellow.svg" alt="Buy Me A Coffee donate button" /></a></span>
<span class="badge-opencollective"><a href="https://opencollective.com/bevry" title="Donate to this project using Open Collective"><img src="https://img.shields.io/badge/open%20collective-donate-yellow.svg" alt="Open Collective donate button" /></a></span>
<span class="badge-crypto"><a href="https://bevry.me/crypto" title="Donate to this project using Cryptocurrency"><img src="https://img.shields.io/badge/crypto-donate-yellow.svg" alt="crypto donate button" /></a></span>
<span class="badge-paypal"><a href="https://bevry.me/paypal" title="Donate to this project using Paypal"><img src="https://img.shields.io/badge/paypal-donate-yellow.svg" alt="PayPal donate button" /></a></span>
<span class="badge-wishlist"><a href="https://bevry.me/wishlist" title="Buy an item on our wishlist for us"><img src="https://img.shields.io/badge/wishlist-donate-yellow.svg" alt="Wishlist browse button" /></a></span>

<h3>Contributors</h3>

These amazing people have contributed code to this project:

<ul><li><a href="http://balupton.com">Benjamin Lupton</a> — <a href="https://github.com/bevry/feedr/commits?author=balupton" title="View the GitHub contributions of Benjamin Lupton on repository bevry/feedr">view contributions</a></li>
<li><a href="https://github.com/Zearin">Zearin</a> — <a href="https://github.com/bevry/feedr/commits?author=Zearin" title="View the GitHub contributions of Zearin on repository bevry/feedr">view contributions</a></li>
<li><a href="http://github.com/apps/dependabot">dependabot[bot]</a> — <a href="https://github.com/bevry/feedr/commits?author=dependabot[bot]" title="View the GitHub contributions of dependabot[bot] on repository bevry/feedr">view contributions</a></li>
<li><a href="http://github.com/apps/dependabot-preview">dependabot-preview[bot]</a> — <a href="https://github.com/bevry/feedr/commits?author=dependabot-preview[bot]" title="View the GitHub contributions of dependabot-preview[bot] on repository bevry/feedr">view contributions</a></li></ul>

<a href="https://github.com/bevry/feedr/blob/master/CONTRIBUTING.md#files">Discover how you can contribute by heading on over to the <code>CONTRIBUTING.md</code> file.</a>

<!-- /BACKERS -->


<!-- LICENSE/ -->

<h2>License</h2>

Unless stated otherwise all works are:

<ul><li>Copyright &copy; 2012+ <a href="http://bevry.me">Bevry Pty Ltd</a></li></ul>

and licensed under:

<ul><li><a href="http://spdx.org/licenses/MIT.html">MIT License</a></li></ul>

<!-- /LICENSE -->


HISTORY.md
# History

## v4.8.0 2019 December 9

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v4.7.0 2019 December 1

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v4.6.0 2019 December 1

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v4.5.0 2019 November 18

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v4.4.0 2019 November 18

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v4.3.0 2019 November 13

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v4.2.0 2019 November 9

-   Updated [base files](https://github.com/bevry/base) and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v4.1.0 2019 November 9

-   Updated [base files](https://github.com/bevry/base) and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v4.0.0 2019 October 2

-   Updated [base files](https://github.com/bevry/base) and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)
-   Changed minimum supported node version from v4 to v8, due to failing tests on old versions
-   Dropped `require('feedr').Feedr` use `require('feedr')` instead

## v3.0.0 2018 January 25

-   Minimum node version is now version 4 due to the nested dependencies [hoek](https://www.npmjs.com/package/hoek) and [hawk](https://www.npmjs.com/package/hawk) publishing code that does not work with earlier node verisons - if only they used [editions](https://github.com/bevry/editions)
-   Updated base files
-   Updated dependencies

## v2.13.5 2017 February 28

-   Internal: swapped a `forof` loop for `forin` loop, for better compat with old environments
-   Added error for `checkResponse` instead of `check`

## v2.13.4 2017 February 28

-   Fixed checks never firing (regression since v2.10.0)
-   Updated dependencies
-   Updated base files

## v2.13.3 2016 June 18

-   Fixed legacy API `require('feedr').Feedr` not existing (regression since v2.13.0)

## v2.13.2 2016 June 13

-   Fixed unknown clear method error (regression since v2.13.0)
-   By default, use all plugins, like v2.12.0 and earlier (regression since v2.13.0) - Thanks to [Chris Krycho](https://github.com/chriskrycho) for [issue #6](https://github.com/bevry/feedr/issues/6)

## v2.13.1 2016 June 8

-   Meta change, correct syntaxes that are used in editions

## v2.13.0 2016 June 8

-   Converted from CoffeeScript to JavaScript
-   Updated dependencies

## v2.12.0 2015 March 17

-   Better YAML parser

## v2.11.1 2015 March 16

-   Updated dependencies

## v2.11.0 2015 February 7

-   Updated dependencies

## v2.10.4 2014 December 12

-   Certain failures will now emit warning messages
-   Updated dependencies

## v2.10.3 2014 August 8

-   Fixed with better tests

## v2.10.2 2014 August 8

-   Fixed specific parsing

## v2.10.1 2014 August 8

-   Fixed xml parsing

## v2.10.0 2014 August 8

-   Large internal refactor
-   `checkResponse` is now renamed to `check`, b/c enabled
-   If a feed fails, the `err` argument will indicate this instead of merely outputting the error
-   Feed details are now updated on the inputted feed details object, instead of a clone
-   Now supports `raw`/`false` parse option

## v2.9.1 2014 August 3

-   Updated dependencies

## v2.9.0 2014 May 21

-   `cache` option now defaults to one day `1000*60*60*24` to avoid superflous requests, refer to the updated readme for what the new values mean
-   `parse` option is now only a boolean
-   parsing auto-detection now checks content-types as well as the previous extension check
-   Updated dependencies

## v2.8.0 2014 May 3

-   Added support for `"preferred"` and millisecond values for `cache` configuration option
-   Updated dependencies

## v2.7.7 2014 January 10

-   Updated dependencies

## v2.7.6 2014 January 2

-   Updated dependencies
-   More debug log messages

## v2.7.5 2013 November 27

-   Use the `Wget/1.14 (linux-gnu)` User-Agent by default
-   Updated dependencies

## v2.7.4 2013 November 13

-   `readFeeds` can now accept default options to apply to each feed that will be ready

## v2.7.3 2013 October 31

-   Can now parse CSON files

## v2.7.2 2013 October 4

-   Fixed `checkResponse` option
-   Better catching of errors with invalid JSON

## v2.7.1 2013 October 3

-   Fixed `TypeError: Cannot call method 'toString' of null` (regression since v2.7.0)

## v2.7.0 2013 October 3

-   Can now parse yaml files
-   Can now customise parsing via `parse` feed option
-   Parsing is now determined by extensions instead of file formats (b/c break) - If your url does not have an extension you must eplicitly set the `parse` feed option

## v2.6.0 2013 October 2

-   Some options have changed (b/c break)
-   Much more intelligent cache handling - We now use the expires and etag headers for caching
-   Added the ability to customise the request options
-   Dependency upgrades

## v2.5.1 2013 June 29

-   Dependency upgrades

## v2.5.0 2013 May 27

-   `cache`, `cacheTime`, `timeout`, `checkResult(data)` can now be specified on the feed details
-   `checkResult(data)` should return an Error instance if it finds a problem with the fetched data
-   `cacheTime` and `checkResult` have default implementations for github

## v2.4.4 2013 April 22


source/index.js
'use strict'

// Requires
const extendr = require('extendr')
const eachr = require('eachr')
const { TaskGroup } = require('taskgroup')
const typeChecker = require('typechecker')
const safefs = require('safefs')
const safeps = require('safeps')
const pathUtil = require('path')
const request = require('request')

// Define
class Feedr {
	// Helpers
	static create(...args) {
		return new Feedr(...args)
	}

	// Check to see if the feed === still relevant
	// feed={cache}, cache=boolean/`preferred`/number
	// metaData={expires, date}
	// return boolean
	static isFeedCacheStillRelevant(feed, metaData) {
		return (
			feed.cache && // User always wants to use cache
			(feed.cache === 'preferred' ||
				// If the cache === still relevant according to the website
				(metaData.expires && new Date() < new Date(metaData.expires)) ||
				// If the cache === still relevant according to the user
				(typeChecker.isNumber(feed.cache) &&
					metaData.date &&
					new Date() <
						new Date(new Date(metaData.date).getTime() + feed.cache)))
		)
	}

	// Constructor
	constructor(config = {}) {
		// Prepare
		const me = this

		// Extend and dereference our configuration
		this.config = extendr.deep(
			{
				log: null,
				cache: 1000 * 60 * 60 * 24, // one day by default
				tmpPath: null,
				requestOptions: null,
				plugins: null
			},
			this.config || {},
			config
		)

		// Get the temp path right away
		safeps.getTmpPath(function(err, tmpPath) {
			if (err) {
				console.error(err)
			} else {
				me.config.tmpPath = tmpPath
			}
		})
	}

	// Log
	log(...args) {
		if (this.config.log) this.config.log(...args)
		return this
	}

	// Read Feeds
	// feeds = {feedName:feed}
	// next(err,result)
	readFeeds(...args) {
		// Prepare
		const me = this
		const failures = []

		// Prepare options
		let feeds = null
		const defaultfeed = {} // what is this?
		let next = null

		// Extract the configuration from the arguments
		args.forEach(function(arg, index) {
			if (typeChecker.isFunction(arg)) {
				next = arg
			} else if (typeChecker.isArray(arg)) {
				feeds = arg
			} else if (typeChecker.isPlainObject(arg)) {
				if (index === 0) {
					feeds = arg
				} else {
					extendr.extend(defaultfeed, arg)
				}
			}
		})

		// Extract
		const results = {}

		// Tasks
		const tasks = TaskGroup.create({
			concurrency: 0,
			abortOnError: false
		}).done(function() {
			let message = 'Feedr finished fetching'
			let err = null

			if (failures.length !== 0) {
				message +=
					`with ${failures.length} failures:\n` +
					failures
						.map(function(i) {
							return i.message
						})
						.join('\n')
				err = new Error(message)
				me.log('warn', err)
			} else {
				me.log('debug', message)
			}

			next(err, results)
		})

		// Feeds
		eachr(feeds, function(feed, index) {
			tasks.addTask(function(complete) {
				// Prepare
				if (typeChecker.isString(feed)) {
					feed = { url: feed }
				}
				feeds[index] = feed = extendr.deep({}, defaultfeed, feed)

				// Read
				me.readFeed(feed, function(err, data) {
					// Handle
					if (err) {
						me.log(
							'warn',
							`Feedr failed to fetch [${feed.url}] to [${feed.path}]`,
							err.stack
						)
						failures.push(err)
					} else {
						results[index] = data
					}

					// Complete
					complete(err)
				})
			})
		})

		// Start
		tasks.run()

		// Chain
		return this
	}

	// Prepare Feed Details
	prepareFeed(feed) {
		// Set defaults
		if (feed.hash == null)
			feed.hash = require('crypto')
				.createHash('md5')
				.update(`feedr-${JSON.stringify(feed.url)}`)
				.digest('hex')
		if (feed.basename == null)
			feed.basename = pathUtil.basename(feed.url.replace(/[?#].*/, ''))
		if (feed.extension == null) feed.extension = pathUtil.extname(feed.basename)
		if (feed.name == null) feed.name = feed.hash + feed.extension
		if (feed.path == null)
			feed.path = pathUtil.join(this.config.tmpPath, feed.name)
		if (feed.metaPath == null)
			feed.metaPath =
				pathUtil.join(this.config.tmpPath, feed.name) + '-meta.json'
		if (feed.cache == null) feed.cache = this.config.cache
		if (feed.parse == null) feed.parse = true
		if (feed.parse === 'raw') feed.parse = false
		if (feed.check == null) feed.check = true
		if (feed.plugins == null)
			feed.plugins = this.config.plugins || 'github xml cson json yaml string'
		if (feed.metaData == null) feed.metaData = {}

		// Return
		return feed
	}

	// Cleanup response data
	cleanData(data) {
		// Prepare
		const me = this
		const keys = []

		// Discover the keys inside data, and delve deeper
		eachr(data, function(value, key) {
			if (typeChecker.isPlainObject(data)) {
				data[key] = me.cleanData(value)
			}
			keys.push(key)
		})

		// Check if we are a simple rest object
		// If so, make it a simple value
		if (keys.length === 1 && keys[0] === '_content') {
			data = data._content
		}

		// Return the result
		return data
	}

	// Read Feed
	// next(err,data)
	readFeed(...args) {
		// Prepare
		const me = this
		let url, feed, next

		// Extract the configuration from the arguments
		args.forEach(function(arg) {
			if (typeChecker.isString(arg)) {
				url = arg
			} else if (typeChecker.isFunction(arg)) {
				next = arg
			} else if (typeChecker.isPlainObject(arg)) {
				feed = arg
			}
		})

		// Check for url
		if (!feed) feed = {}
		if (url) feed.url = url
		if (!feed.url) {
			next(new Error('Feed url was not supplied'))
			return this
		}

		// Check deprecations
		if (feed.checkReponse) {
			next(new Error('Feed checkResponse option is deprecated for check'))
			return this
		}

		// Ensure optional
		feed = this.prepareFeed(feed)

		// Plugins
		const plugins = {}
		if (typeChecker.isString(feed.plugins)) {
			feed.plugins = feed.plugins.split(' ')
		}
		if (typeChecker.isArray(feed.plugins)) {
			for (let i = 0; i < feed.plugins.length; ++i) {
				const name = feed.plugins[i]
				try {
					plugins[name] = require('./plugins/' + name)
				} catch (err) {
					next(err)
					return this
				}
			}
		}

		// Generators
		function generateParser(name, method, opts, complete) {
			me.log('debug', `Feedr parse [${feed.url}] with ${name} attempt`)
			method(opts, function(err, data) {
				if (err) {
					complete(err)
					return
				}
				if (data) {
					me.log(
						'debug',
						`Feedr parse [${feed.url}] with ${name} attempt, used`
					)
					opts.data = data
				} else {
					me.log(
						'debug',
						`Feedr parse [${feed.url}] with ${name} attempt, ignored`
					)
				}
				complete(null, data)
			})
		}
		function generateChecker(name, method, opts, complete) {
			me.log('debug', `Feedr check [${feed.url}] with ${name} attempt`)
			method(opts, function(err, data) {
				if (err) {
					complete(err)
					return
				}
				me.log(
					'debug',
					`Feedr check [${feed.url}] with ${name} attempt, success`
				)
				complete(null, data)
			})
		}

		// ------------------------------
		// Parser

		let parseResponse = null

		// Specific
		if (typeChecker.isString(feed.parse)) {
			// Exists
			if (
				typeChecker.isFunction(plugins[feed.parse] && plugins[feed.parse].parse)
			) {
				parseResponse = generateParser.bind(
					null,
					feed.parse,
					plugins[feed.parse].parse
				)
			}

			// Missing
			else {
				next(new Error('Invalid parse value: ' + feed.parse))
				return this
			}
		}

		// Custom
		else if (typeChecker.isFunction(feed.parse)) {
			parseResponse = generateParser.bind(null, 'custom', feed.parse)
		}

		// Auto
		else if (feed.parse === true) {
			parseResponse = function(opts, parseComplete) {
				const checkTasks = new TaskGroup().done(parseComplete)
				eachr(plugins, function(value, key) {
					if (value.parse != null) {
						checkTasks.addTask(function(parseTaskComplete) {
							generateParser.bind(
								null,
								key,
								value.parse
							)(opts, function(err, data) {
								if (data) {
									checkTasks.clear()
								}
								parseTaskComplete(err)
							})
						})
					}
				})
				checkTasks.run()
			}
		}

		// Raw
		else {
			parseResponse = function(opts, parseComplete) {
				parseComplete()
			}
		}

		// ------------------------------
		// Checker

		let checkResponse = null

		// Specific
		if (typeChecker.isString(feed.check)) {
			// Exists
			if (
				typeChecker.isFunction(plugins[feed.check] && plugins[feed.check].check)
			) {
				checkResponse = generateChecker.bind(
					null,
					feed.check,
					plugins[feed.check].check
				)
			}

			// Missing
			else {
				next(new Error('Invalid check value: ' + feed.check))
				return this
			}
		}

		// Custom
		else if (typeChecker.isFunction(feed.check)) {
			checkResponse = generateChecker.bind(null, 'custom', feed.check)
		}

		// Auto
		else if (feed.check) {
			checkResponse = function(opts, checkComplete) {
				const checkTasks = new TaskGroup().done(checkComplete)
				eachr(plugins, function(value, key) {
					if (value.check != null) {
						checkTasks.addTask(function(checkTaskComplete) {
							generateChecker.bind(
								null,
								key,
								value.check
							)(opts, checkTaskComplete)
						})
					}
				})
				checkTasks.run()
			}
		}

		// Raw
		else {
			checkResponse = function(opts, checkComplete) {
				checkComplete()
			}
		}

		// Request options
		const requestOptions = extendr.deep(
			{
				url: feed.url,
				timeout: 1 * 60 * 1000,
				encoding: null,
				headers: {
					'User-Agent': 'Wget/1.14 (linux-gnu)'
				}
			},
			me.config.requestOptions || {},
			feed.requestOptions || {}
		)

		// Read a file
		function readFile(path, readFileComplete) {
			// Log
			me.log(
				'debug',
				`Feedr === reading [${feed.url}] on [${path}], checking exists`
			)

			// Check the the file exists
			safefs.exists(path, function(exists) {
				// Check it exists
				if (!exists) {
					// Log
					me.log(
						'debug',
						`Feedr === reading [${feed.url}] on [${path}], it doesn't exist`
					)

					// Exit
					readFileComplete()
					return
				}

				// Log
				me.log(
					'debug',
					`Feedr === reading [${feed.url}] on [${path}], it exists, now reading`
				)

				// It does exist, so let's continue to read the cached fie
				safefs.readFile(path, null, function(err, rawData) {
					// Check
					if (err) {
						// Log
						me.log(
							'debug',
							`Feedr === reading [${feed.url}] on [${path}], it exists, read failed`,
							err.stack
						)

						// Exit
						readFileComplete(err)
						return
					}

					// Log
					me.log(
						'debug',
						`Feedr === reading [${feed.url}] on [${path}], it exists, read completed`
					)

					// Return the parsed cached data
					readFileComplete(null, rawData)
				})
			})
		}

		// Parse a file
		function readMetaFile(path, readMetaFileComplete) {
			// Log
			me.log('debug', `Feedr === parsing meta file [${feed.url}] on [${path}]`)

			// Parse
			readFile(path, function(err, rawData) {
				// Check
				if (err || !rawData) {
					// Log
					me.log(
						'debug',
						`Feedr === parsing meta file [${feed.url}] on [${path}], read failed`,
						err && err.stack
					)

					// Exit
					readMetaFileComplete(err)
					return
				}

				// Attempt
				let data = null
				try {
					data = JSON.parse(rawData.toString())
				} catch (err) {
					// Log
					me.log(
						'warn',
						`Feedr === parsing meta file [${feed.url}] on [${path}], parse failed`,
						err.stack
					)

					// Exit
					readMetaFileComplete(err)
					return
				}

				// Log
				me.log(
					'debug',
					`Feedr === parsing meta file [${feed.url}] on [${path}], parse completed`
				)

				// Exit
				readMetaFileComplete(null, data)
			})
		}

		// Write the feed
		function writeFeed(response, data, writeFeedComplete) {
			// Log
			me.log('debug', `Feedr === writing [${feed.url}] to [${feed.path}]`)

			// Prepare
			const writeTasks = TaskGroup.create({ concurrency: 0 }).done(function(
				err
			) {
				if (err) {
					// Log
					me.log(
						'warn',
						`Feedr === writing [${feed.url}] to [${feed.path}], write failed`,
						err.stack
					)

					// Exit
					writeFeedComplete(err)
					return
				}

				// Log
				me.log(
					'debug',
					`Feedr === writing [${feed.url}] to [${feed.path}], write completed`
				)

				// Exit
				writeFeedComplete(null, data)
			})

			writeTasks.addTask('store the meta data in a cache somewhere', function(
				writeTaskComplete
			) {
				const writeData = JSON.stringify(
					{
						headers: response.headers,
						parse: feed.parse
					},
					null,
					'  '
				)
				safefs.writeFile(feed.metaPath, writeData, writeTaskComplete)
			})

			writeTasks.addTask('store the parsed data in a cache somewhere', function(
				writeTaskComplete
			) {
				const writeData = feed.parse ? JSON.stringify(data) : data
				safefs.writeFile(feed.path, writeData, writeTaskComplete)
			})

			// Fire the write tasks
			writeTasks.run()
		}

		// Get the file via reading the cached copy
		// next(err, data, meta)
		function viaCache(viaCacheComplete) {
			// Log
			me.log('debug', `Feedr === remembering [${feed.url}] from cache`)

			// Prepare
			let meta = null
			let data = null
			const readTasks = TaskGroup.create().done(function(err) {
				viaCacheComplete(err, data, meta && meta.headers)
			})

			readTasks.addTask('read the meta data in a cache somewhere', function(
				viaCacheTaskComplete
			) {
				readMetaFile(feed.metaPath, function(err, result) {
					if (err || !result) {
						viaCacheTaskComplete(err)
						return
					}
					meta = result
					viaCacheTaskComplete()
				})
			})

			readTasks.addTask('read the parsed data in a cache somewhere', function(
				viaCacheTaskComplete
			) {
				readFile(feed.path, function(err, rawData) {
					if (err || !rawData) {
						viaCacheTaskComplete(err)
						return
					}
					if (
						feed.parse === false ||
						(feed.parse === true && meta.parse === false)
					) {
						data = rawData
					} else {
						try {
							data = JSON.parse(rawData.toString())
						} catch (err) {
							viaCacheTaskComplete(err)
							return
						}
					}
					viaCacheTaskComplete()
				})
			})

			// Fire the write tasks
			readTasks.run()
		}

		// Get the file via performing a fresh request
		// next(err, data, meta)
		function viaRequest(viaRequestComplete) {
			// Log
			me.log(
				'debug',
				`Feedr === fetching [${feed.url}] to [${feed.path}], requesting`
			)

			// Add etag if we have it
			if (feed.cache && feed.metaData.etag) {
				if (requestOptions.headers['If-None-Match'] == null) {
					requestOptions.headers['If-None-Match'] = feed.metaData.etag
				}
			}

			// Fetch and Save
			request(requestOptions, function(err, response, data) {
				// Log
				const opts = { feedr: me, feed, response, data }
				me.log(
					'debug',
					`Feedr === fetching [${feed.url}] to [${feed.path}], requested`
				)

				// What should happen if an error occurs
				function handleError(err) {
					// Log
					me.log(
						'warn',
						`Feedr === fetching [${feed.url}] to [${feed.path}], failed`,
						err.stack
					)

					// Exit
					if (feed.cache) {
						viaCache(next)
						return
					}
					viaRequestComplete(err, opts.data, requestOptions.headers)
				}

				// Check error
				if (err) {
					handleError(err)
					return
				}

				// Check cache
				if (feed.cache && response.statusCode === 304) {
					viaCache(next)
					return
				}

				// Determine Parse Type
				parseResponse(opts, function(err) {
					if (err) {
						handleError(err)
						return
					}

					// Log
					me.log(
						'debug',
						`Feedr === fetching [${feed.url}] to [${feed.path}], requested, checking`
					)

					// Exit
					checkResponse(opts, function(err) {
						if (err) {
							handleError(err)
							return
						}
						writeFeed(response, opts.data, function(err) {
							viaRequestComplete(err, opts.data, requestOptions.headers)
						})
					})
				})
			})
		}

		// Refresh if we don't want to use the cache
		if (feed.cache === false) {
			viaRequest(next)
			return this
		}

		// Fetch the latest cache data to check if it === still valid
		readMetaFile(feed.metaPath, function(err, metaData) {
			// There isn't a cache file
			if (err || !metaData) {
				viaRequest(next)
				return
			}

			// Apply to the feed details
			feed.metaData = metaData

			// There === an expires header and it === still valid
			// cache preferred, use cache if exists, otherwise fall back to relevant
			// cache number, use cache if within number, otherwise fall back to relevant
			if (Feedr.isFeedCacheStillRelevant(feed, metaData)) {
				viaCache(next)
				return
			}

			// There was no expires header
			viaRequest(next)
		})

		// Chain
		return this
	}
}

// Exports
module.exports = Feedr
